### PR TITLE
DTP-711: Handle SequenceId edge cases

### PR DESCRIPTION
**Pull request by kaschula**
This PR adds tests and validation handling for the sequenceId return from the configured sync function.

Conditions:
- [x] 1. SequenceId is `undefined`
- [x] 2. SequenceId is `'0'` and there is history in the channel. In which case history should be applied.

Previous PR: https://github.com/ably-labs/models/pull/162, I had to change the branch name because I used the wrong jira reference. Old comments are here are on the original PR

# Comments

In condition 1 where the sequenceId is `undefined`, the solution in this PR is throw an error and set the model state to `errored`. Additionally the `retryable()` logic has been updated to not run the callback function if the model is in the `errored` state. Additionally if we catch an error while in the errored state then we re throw it, propagating it up the call stack. This is how the new `sequenceId is undefined error` will be surfaced to the user. 

# Other changes

Error handling has been improved so that all errors produced by the SDK during event handling function or in places where a user would not be able to catch them have been silenced and are exposed through the 'errored' event.

The docs around the Models's sync and subscribe methods have been updated so that the user knows to catch any errors thrown by these functions.

An integration test has been added that covers all message history being applied to model with sequenceId 0 and, seek and applying history with a sequenceId of 1 or more.

The `uuid` package has been added as a dev dependancy. 

The current integration test located in `./src/Model.integration.test.ts` has been wrapped in its own describe as the test set up new test is different. This test are now isolated. I have also added uuids to the channel name so that each test run uses a different channel.

### CODE REVIEW

File: src/stream/Middleware.test.ts [hunk 1]
```diff
@@ -266,25 +266,26 @@ describe('OrderedHistoryResumer', () => {
...
       createMessage(3),
       createMessage(2),
-      createMessage(1),
     ];
     const page1 = history;
     const page2 = [];
 
     expect(middleware.addHistoricalMessages(shuffle(page1))).toBe(false);
-    expect(middleware.addHistoricalMessages(shuffle(page2))).toBe(true);
+    expect(middleware.addHistoricalMessages(page2)).toBe(true);
```
Comment (mschristensen) [inline src/stream/Middleware.test.ts:286]: is there a reason to remove the shuffle here?

**Review by mschristensen — APPROVED** [review]
Looks good, minor comments then happy to merge :)

File: src/stream/Middleware.test.ts (hunk 1 above)
Comment (kaschula) [reply src/stream/Middleware.test.ts:286]: I removed it because `page2` is an empty array so there was nothing to shuffle. Would you like me to put it back?

**Comment by mschristensen** [comment]
Let's merge https://github.com/ably-labs/models/pull/164 into this and then happy to approve :)
